beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -1725,13 +1733,63 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         result
     }
 
+    fn prune_abandoned_forks(
+        &self,
+        old_finalized_block: &Hash256,
+        new_finalized_block: &Hash256,
+    ) -> Result<(), Error> {
+        let old_finalized_slot = self
+            .get_block(old_finalized_block)?
+            .ok_or_else(|| Error::MissingBeaconState(*old_finalized_block))?
+            .message
+            .slot;
+
+        let new_finalized_slot = self
+            .get_block(new_finalized_block)?
+            .ok_or_else(|| Error::MissingBeaconState(*new_finalized_block))?
+            .message
+            .slot;
+
+        // Collect hashes from new_finalized_block up to old_finalized_block
+        let newly_finalized_blocks: HashSet<Hash256> = HashSet::from_iter(
+            ParentRootBlockIterator::new(&*self.store, *new_finalized_block)
+                .take_while(|(block_hash, _)| block_hash != old_finalized_block)
+                .map(|(block_hash, _)| block_hash),
+        );
+
+        // Forget blocks with the slot number in the range [old_finalized_slot, new_finalized_slot]
+        // that are not in the newly_finalized_blocks.
+        for (head, _) in self.heads() {
```

Comment:
Concern: I don't see anywhere in code `self.head_tracker` being shrinked.  It only grows.  Isn't this a problem in very long running nodes?

Reply:
Ah yes indeed, this is something I thought we would pick up once we started pruning. I had forgotten about that.

Perhaps we should shrink the `head_tracker` now we've deleted the blocks it points to?

Reply:
Would you mind if I did that in a separate PR?  I have some tiny test harness refactoring that I want to attempt too.